src/app/shared/form/builder/ds-dynamic-form-ui/models/autocomplete/dynamic-autocomplete.component.ts

```diff
@@ -0,0 +1,296 @@
+        return DynamicAutocompleteService.pretifyFundingSuggestion(complexInputList[1], complexInputList[2]);
+      } else if (suggestion instanceof ExternalSourceEntry) {
+        // suggestion from the openAIRE
+        const fundingProjectCode = this.getProjectCodeFromId(suggestion.id);
+        const fundingName = suggestion.metadata['project.funder.name'][0].value;
+        return DynamicAutocompleteService.pretifyFundingSuggestion(fundingProjectCode, fundingName);
+      }
+    }
+    // @ts-ignore
+    return suggestion.display;
+  }
+
+  /**
+   * Converts a text values stream from the `<input>` element to the array stream of the items
+   * and display them in the typeahead popup.
+   */
+  search = (text$: Observable<string>) =>
+    text$.pipe(
+      debounceTime(300),
+      distinctUntilChanged(),
+      tap(() => this.changeSearchingStatus(true)),
+      switchMap((term) => {
+        // min 3 characters
+        if (term === '' || term.length < this.model.minChars) {
+          return observableOf({ list: [] });
+        } else {
+          let response: Observable<PaginatedList<ExternalSourceEntry | MetadataValue>>;
+          if (this.isSponsorInputType) {
+            // if openAIRE
+            // @ts-ignore
+            const fundingType = this.model.parent.group[0].value;
+            if (isNotEmpty(fundingType) && ['euFunds', 'EU'].includes(fundingType.value)) {
+              // eu funding
+              response = this.lookupRelationService.getExternalResults(
+                this.getOpenAireExternalSource(), this.getFundingRequestOptions(term));
+            } else {
+              // non eu funding
+              response = this.metadataValueService.findByMetadataNameAndByValue(SPONSOR_METADATA_NAME, term);
+            }
+          } else {
+            // metadataValue request
+            response = this.metadataValueService.findByMetadataNameAndByValue(this.model.name, term);
+          }
+          return response.pipe(
+            tap(() => this.searchFailed = false),
+            catchError((error) => {
+              this.searchFailed = true;
+              return observableOf(buildPaginatedList(
+                new PageInfo(),
+                []
+              ));
+            }));
+        }
+      }),
+      map((list: any) => {
+        return list.page;
+      }),
+      tap(() => this.changeSearchingStatus(false)),
+      merge(this.hideSearchingWhenUnsubscribed))
+
+
+  /**
+   * Only for the local.sponsor complex input type
+   * The external funding is composed as one complex input field
+   * @param updateValue external funding from the openAIRE
+   */
+  composeSponsorInput(updateValue) {
+    // set prefix to distinguish composed complex input in the complex.model.ts - get method
+    let newValue = AUTOCOMPLETE_COMPLEX_PREFIX + SEPARATOR;
+    let fundingType = 'N/A';
+    let fundingProjectCode = '';
+
+    if (updateValue.id.startsWith(EU_PROJECT_PREFIX)) {
+      fundingType = 'EU';
```

Comment:
either it is a generic component or very specific

Reply:
I created DsDynamicSponsorAutocomplete component which extends DsDynamicAutocomplete